File: ios/VisionCameraViewManager.m
#import "React/RCTViewManager.h"

@interface RCT_EXTERN_MODULE(VisionCameraViewManager, RCTViewManager)

RCT_EXPORT_VIEW_PROPERTY(onCapture, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onError, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onRecognitionUpdate, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onSharpnessScoreUpdate, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onBarcodeDetected, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onBoundingBoxesUpdate, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(enableFlash, BOOL)
RCT_EXPORT_VIEW_PROPERTY(zoomLevel, NSNumber)
RCT_EXPORT_VIEW_PROPERTY(scanMode, NSString)
RCT_EXPORT_VIEW_PROPERTY(autoCapture, BOOL)
RCT_EXPORT_VIEW_PROPERTY(scanArea, NSDictionary)
RCT_EXPORT_VIEW_PROPERTY(detectionConfig, NSDictionary)
RCT_EXPORT_VIEW_PROPERTY(frameSkip, NSNumber)

RCT_EXTERN_METHOD(capture:(nonnull NSNumber *)node)
RCT_EXTERN_METHOD(stop:(nonnull NSNumber *)node)
RCT_EXTERN_METHOD(start:(nonnull NSNumber *)node)
RCT_EXTERN_METHOD(toggleFlash:(nonnull NSNumber *)node enabled:(BOOL)enabled)
RCT_EXTERN_METHOD(setZoom:(nonnull NSNumber *)node level:(nonnull NSNumber *)level)

@end

File: ios/VisionCameraViewManager.swift
import Foundation
import AVFoundation

@available(iOS 13.0, *)
@objc(VisionCameraViewManager)
class VisionCameraViewManager: RCTViewManager {

    @objc func getComponent(_ node: NSNumber, completion: @escaping (RNVisionCameraView?) -> Void) {
        DispatchQueue.main.async {
            if let component = self.bridge.uiManager.view(forReactTag: node) as? RNVisionCameraView {
                completion(component)
            } else {
                completion(nil)
            }
        }
    }

    @objc func capture(_ node: NSNumber) {
        getComponent(node) { component in
            component?.capture()
        }
    }

    @objc func stop(_ node: NSNumber) {
        getComponent(node) { component in
            component?.stop()
        }
    }

    @objc func start(_ node: NSNumber) {
        getComponent(node) { component in
            component?.start()
        }
    }

    @objc func toggleFlash(_ node: NSNumber, enabled: Bool) {
        getComponent(node) { component in
            component?.toggleFlash(enabled: enabled)
        }
    }

    @objc func setZoom(_ node: NSNumber, level: NSNumber) {
        getComponent(node) { component in
            component?.setZoom(level: CGFloat(level.floatValue))
        }
    }

    override func view() -> UIView! {
        return RNVisionCameraView()
    }

    override static func requiresMainQueueSetup() -> Bool {
        return true
    }
}
